### PR TITLE
Comlink wrapper improvements, more Sheaf bugfixes, performance boosts

### PR DESCRIPTION
**Pull request by Monkatraz**
Does what it says on the tin, basically.

In detail:
* Added timeouts for bound functions on `AbstractWorkerBase`
* Added the ability to provide fallback values for `AbstractWorkerBase`
* Limited bound functions to methods and classes for `AbstractWorkerBase`
* Added checks for `matchMedia` where ever it's used
* Added a new `wordCount` function to `ftml-wasm`, which uses the FTML AST to count the number of visible words
* Got rid of the `RenderHandler` class and the `Wikitext` component, replacing them with the faster and simpler `MorphingWikitext` component
* Added the `encode` and `decode` utility functions to `@wikijump/util`
* Made it so that the `Timeout` class has an _optional_ callback

### CODE REVIEW

File: web/modules/comlink/src/index.ts [hunk 1]
```diff
@@ -17,62 +22,112 @@ export type {
...
+          throw new Error(`Method "${method}" timed out!`)
         }
       }
     }
+
+    return Derived
+  }
+
+  /** Tries to run a default method if the worker couldn't be started. */
+  private async _baseTryToGetDefault(method: keyof Methods<T>, args: any[]) {
```
Comment (emmiegit) [inline web/modules/comlink/src/index.ts:118]: Why is this needed?

File: web/modules/sheaf/src/components/MorphingWikitext.svelte [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    const result =
+      typeof wikitext === "string"
+        ? await FTML.renderHTML(wikitext, undefined, "draft")
+        : await wikitext.render()
+
+    timeCompile = measureCompile()
+    return result
+  })
+
+  const patch = createLock(async ({ html, styles }: RenderedHTML) => {
```
Comment (emmiegit) [inline web/modules/sheaf/src/components/MorphingWikitext.svelte:58]: For my information, what does the `createLock` utility do?

File: web/modules/sheaf/src/state.ts [hunk 3]
```diff
@@ -27,30 +53,121 @@ export class SheafState {
...
+  /** Retrieves the value of the editor as an `Uint8Array`. */
+  @Memoize()
   async buffer() {
     return await textBuffer(this.doc)
   }
 
+  /** Renders the editor with FTML. */
+  @Memoize()
+  async render() {
+    return await FTML.renderHTML(await this.value(), undefined, "draft")
```
Comment (emmiegit) [inline web/modules/sheaf/src/state.ts:71]: Are these methods memoized because the value doesn't change, or is Typescript invalidating the cached result if any of the fields update?

**Review by emmiegit — APPROVED** [review]

File: web/modules/comlink/src/index.ts (hunk 1 above)
Comment (Monkatraz) [reply web/modules/comlink/src/index.ts:118]: I wanted an easy way to provide fallback values if the worker can't start. It's a bit messy to do a override function and check for this yourself, so instead I provided a specific object that you can use for providing fallbacks. This function tries to run/retrieve a fallback value for a worker method, if it exists.

File: web/modules/sheaf/src/components/MorphingWikitext.svelte (hunk 2 above)
Comment (Monkatraz) [reply web/modules/sheaf/src/components/MorphingWikitext.svelte:58]: Spits out a function that when called multiple times, will wait for each previous call to finish. Basically, only one instance of the function can be active at a time.

File: web/modules/sheaf/src/state.ts (hunk 3 above)
Comment (Monkatraz) [reply web/modules/sheaf/src/state.ts:71]: The `SheafState` class is a immutable container basically. A new one is made when the editor updates. The idea with these functions being memoized is that if the editor state isn't changing but you're still requesting stuff from it, it won't do any unnecessary work.

An example of this is when you switch around the different preview tabs, like Result, HTML, back to Result, etc. It'll just use the previous rendered result.

**Comment by Monkatraz** [comment]
thanks @ammongit
